### ISSUE TITLE
Incorrect endings for street names in narratives in Russian.

### ISSUE THREAD

**Issue by molind** [valhalla/valhalla#731]
Maybe you could suggest me how is should fix cases for nouns. E.g. "улица", but "по улиц**е**", "на улиц**у**". Word ending changes in different cases. There are two ways to fix this issue:

- just replace "улица", "проспект", "переулок" and other common names to appropriate case for specific narrative. It's easiest and few lines of code solution, but it solves only 80%-95% of issues with incorrect endings.
- add structural parsing of Russian words to correctly handle endings in street names. E.g. "Красная улица", but "по Красн**ой** улиц**е**", "на Красн**ую** улиц**у**".  It could cause one more dependency to odin to structural parse street names and change them for different cases.

**Comment by kevinkreiser** [comment]
i like the latter approach. other languages have this issue as well so if we could put in something like the pluralization that could be overridden i think that would be the best.

just to be sure i understand exactly what the issue is (my russian knowledge is almost zero)...

are you saying that the preposition `по` vs `на` changes the ending of the noun `улица`? german has something similar where the preposition changes the article used with the noun depending on the preposition used and the gender of the noun itself, so it could benefit from this as well. romance languages also have this concept wrt prepositions and gendered articles.

**Comment by molind** [comment]
> are you saying that the preposition по vs на changes the ending of the noun улица?

Yes. When you see **по** it's prepositional case, **на** - accusative case. Case depends from a verb and it changes ending in noun and related adjectives.

**Comment by molind** [comment]
I'll check available libraries for structural text analysis.

**Comment by kevinkreiser** [comment]
i was thinking of something custom, that you could modify the text directly.

so if the line were:

`... preposition some_name some_street_type ...`

maybe the function could find the preposition next to the street name replacement tag and it would get the street its using to do the replace. then you could do the replacement and modify the street name string before doign the replacement according to if you found the right preposition or not. in other languages we'd want to also be able to modify the preposition as well which is fine if what is returned is just the finally full string with the street name tag replaced.

is any of that making sense? you'd need a map of prepositions and common street types for a given languages overridden method.

**Comment by kevinkreiser** [comment]
@dgearhart can talk more to the exact details of the code

**Comment by molind** [comment]
Word order in Russian could be different from `... preposition some_name some_street_type ...`. But it makes sense. By using preposition we could detect case. And by case, we could modify ending in street name and street type.

**Comment by kevinkreiser** [comment]
yep i wasnt implying any assumptions about word order. the part in the overridden method that detects the preposition would be responsible for knowing where to look for the preposition. i was just using english as a means of conveying the point using a language that is mutually intelligible to us :smile:

**Comment by dgearhart** [comment]
here is where the street names are formed:
https://github.com/valhalla/valhalla/blob/master/src/odin/narrativebuilder.cc#L3739-L3763

depending on scope of detection and replacement would impact where logic makes sense
I am not sure if you need to create Russian version of street name class - but that is possible.
We have US specific items when we form verbal strings. The verbal strings are formatted to work better with TTS engines

The word order is usually handled for each language by the phrases in the corresponding language json file

**Comment by mlehtima** [comment]
Similar feature would be needed also for Finnish street names because Finnish language has very limited number of prepositions and in most cases grammatical cases are used instead to achieve the same effect.
